File: Reschedule Meetings for Maximum Free Time I/Reschedule Meetings for Maximum Free Time I.py
# Solution for Reschedule Meetings for Maximum Free Time I in PY

class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: list[int], endTime: list[int]) -> int:
        n = len(startTime)
        freeGaps = []

        freeGaps.append(startTime[0])
        for i in range(1, n):
            freeGaps.append(startTime[i] - endTime[i - 1])
        freeGaps.append(eventTime - endTime[-1])

        maxFree = 0
        currSum = 0
        left = 0

        for right in range(len(freeGaps)):
            currSum += freeGaps[right]

            while right - left + 1 > k + 1:
                currSum -= freeGaps[left]
                left += 1

            maxFree = max(maxFree, currSum)

        return maxFree

File: Reschedule Meetings for Maximum Free Time I/Reschedule Meetings for Maximum Free Time I.cpp
// Solution for Reschedule Meetings for Maximum Free Time I in CPP

class Solution {
public:
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        int n = startTime.size();

        // Step 1: Calculate all free time gaps
        vector<int> freeGaps;

        // Initial free time before first meeting
        freeGaps.push_back(startTime[0]);

        // Free time between consecutive meetings
        for (int i = 1; i < n; i++) {
            freeGaps.push_back(startTime[i] - endTime[i - 1]);
        }

        // Free time after the last meeting till event ends
        freeGaps.push_back(eventTime - endTime[n - 1]);

        // Step 2: Use sliding window to find max sum of any (k + 1) free gaps
        int maxFree = 0;
        int currSum = 0;
        int left = 0;

        for (int right = 0; right < freeGaps.size(); right++) {
            currSum += freeGaps[right];

            // Maintain window size of (k + 1) gaps
            while (right - left + 1 > k + 1) {
                currSum -= freeGaps[left];
                left++;
            }

            // Update max free time
            maxFree = max(maxFree, currSum);
        }

        return maxFree;
    }
};
